docs/notes/05-linters-and-tests.md

```diff
@@ -0,0 +1,193 @@
+
+### Additional Resources
+
+The topic of function testing is broad and it is an active area of research. Here are some potentially interesting resources:
+
+#### Websites
+1. https://www.fuzzingbook.org/#A-Textbook-for-Paper,-Screen,-and-Keyboard
+2. https://www.softwaretestinghelp.com/
+
+#### Books
+1. ["Clean Code: A Handbook of Agile Software Craftsmanship" by Robert C. Martin](https://www.goodreads.com/book/show/3735293-clean-code)
+2. ["Test Driven Development: By Example" by Kent Beck](https://www.goodreads.com/book/show/387190.Test_Driven_Development)
+
+## Running Tests and Linters with Tox
+
+We recommend using [tox](https://tox.wiki/en/latest/) to run tests and linters. Tox is a command-line tool that automates and simplifies testing and development workflows in Python projects. It provides a consistent and reproducible environment for testing and running code across multiple configurations.
+
+### Benefits of Using Tox
+
+- **Reproducible Testing**: Tox enables developers to define and run tests consistently across different platforms and environments. This ensures that tests produce consistent results and helps catch potential issues that might arise on specific configurations.
+
+- **Continuous Integration (CI) Support**: Tox is commonly used in CI systems, such as Jenkins, Travis CI, and GitHub Actions. It facilitates easy integration with CI pipelines, allowing for automated testing and continuous delivery.
+
+- **Dependency Management**: By encapsulating dependencies within each test environment, Tox helps manage and isolate dependencies, reducing conflicts and ensuring that tests run in the intended environment.
+
+- **Code Quality and Coverage**: Tox can be configured to run additional tools like code linters, static analyzers, and test coverage tools. This helps maintain code quality and ensure sufficient test coverage.
+
+### Usage
+
+#### Installation
+
+Install `tox` from pypi: `pip install tox`.
+
+#### Configuration
+
+To configure Tox, edit the tox.ini file located in the main repository directory. You can use the provided tox.ini 
+file in this template repository as a starting point for setting up your tests and linters.
+
+Here is a description of the initialization settings:
+
+##### [tox] section
+
+- `isolated_build = True`: Enables isolated build mode, which creates separate virtual environments for each test environment. This helps ensure clean and isolated testing environments.
+- `envlist = env1,env2,...`: Specifies the names of test environments to be executed.
+
+##### [testenv] section
+
+- `extras = dev`: The tox environment will be installed with `pip install .[dev]`, installing the optional `dev` dependencies specified in `pyproject.toml` section `[project.optional-dependencies]`. By using this (instead of `deps`), we allow the individual tests / linters to be run either separately or within tox and avoid having to list them in both `tox.ini` and `pyproject.toml`.
```

Comment:
```suggestion
- `extras = dev`: The tox environment will be installed with `pip install .[dev]`, installing the optional `dev` dependencies specified in `pyproject.toml` section `[project.optional-dependencies]`. By using this (instead of `deps`), we allow the individual tests / linters to be run either separately or within tox and avoid having to list them in both `tox.ini` and `pyproject.toml`. See the [packaging note](02-packaging.md) for some more details on optional dependencies.
```